### PR TITLE
Fix for id calculation

### PR DESCRIPTION
**Pull request by fix**
This is the fix for ill calculation of id (80d75c7b90288246199e4a97ba726bad6639595ef92ad7c2bd14fd31563241ab)

Tested on mainnet full rebuild

### CODE REVIEW

**Comment by fix** [comment]
@faustbrian i dont understand what you mean.

Essentially the issue is v1 mainnet had issue on tx type 1 and 4 where recipientId (which is useless) was an optional field. Some tx have it some have not. It also have issue on verification of the signature

The idea for the code is that if i get those tx and verification failed, we try to set or remove this field and try to verify again signature and recompute id.

**Comment by fix** [comment]
ah ok gotcha what you mean.

**Comment by fix** [comment]
Can you find devnet tx that had issues?

**Comment by fix** [comment]
At some point we should put those examples in tests to be sure we dont have regressions on them

**Comment by fix** [comment]
also thinking loudly about some refactoring: Instead of shenanigans to be able to fix v1 serialisations issue mixed with aip11 implementation, what about (2 options):
- having a function passing deserialisation AND original transaction and performing the the needed fix. pro: no issue about injection attack but difficulties to sort out how legit is the original data
- having a function when receiving a new tx (from blocks or single) and perform the analysis. pro: less redundancy, more prone to injection.

**Comment by alexbarnsley** [comment]
@fix I briefly started but Brian wanted to hold off on it, here: https://github.com/ArkEcosystem/core/pull/790

You can find the exact txs here:

https://github.com/ArkEcosystem/core/issues/754
https://github.com/ArkEcosystem/core/issues/715

I think this is what Brian is referring to

**Comment by fix** [comment]
test passing on models

**Comment by faustbrian** [comment]
As @alexbarnsley already linked, in https://github.com/ArkEcosystem/core/issues/754 are 3 transactions from the old devnet that deserialised as type 1.

They were initially signed **with** a recipient ID, which shouldn't be present on type 1, so their IDs are only valid with a recipient ID **before** computing the transaction ID.

All other type 1 transactions on the old devnet were signed **without** a recipient ID so their IDs are only valid if the recipient ID is added **after** computing the transaction ID.

We would basically need to check for type 1 and 4 if they already have a recipient ID which means we need to compute the transaction ID **with** the recipient ID included, if not we need to add it after the transaction ID is computed.

**Comment by fix** [comment]
added those tx on tests

**Comment by faustbrian** [comment]
Those 3 transactions will be **valid with** the recipient address but all others failed on the old devnet.

I will ask tharude if he still has a v1 devnet snapshot that we could use to reproduce the issue with the IDs.

1. sign **type 1** transaction **without** recipient ID
2. **deserialise** it with @arkecosystem/crypto __(it will add the recipient ID)__
3. transaction ID will be different __(because step 2 adds the recipient ID which results in a different byte representation)__

1. sign **type 1** transaction **without** recipient ID
2. modify the **deseralize** method to add the recipient ID **AFTER** computing the ID
3. deserialise it with **@arkecosystem/crypto** __(it will add the recipient ID)__
4. transaction ID will be the same as before deserialising

**Comment by faustbrian** [comment]
Adrian will restore a v1 snapshot that we can use to query the devnet transactions.

**Comment by fix** [comment]
this branch fix this issue. The way i handle it is the following:
- serialize the tx
- deserialize the tx
- if v1 apply compatibility fix
- if verification failed and type is 1 or 4, remove recipientId and recompute

**Comment by fix** [comment]
the last 2 steps are handled here:
https://github.com/ArkEcosystem/core/pull/956/files#diff-20020cdc4aa37467285162d1bf6da2e0R80

**Comment by fix** [comment]
maybe one clearer solution would be to handle different serialization if v1 or v2, at the price of duplicated code

**Comment by fix** [comment]
ie 
```
if tx.version = 1 -> serializeV1(tx)
else serializeAIP11(tx)
```

**Comment by fix** [comment]
well that is likely better actually, because the way this is coded we could build the new AIP11 x (resignation etc...) the v1 way...

**Comment by fix** [comment]
ok i have finally finished the work on this. Just needs now to fix a few tests and we are good to merge

File: packages/core/lib/config/mainnet/plugins.js [hunk 1]
```diff
@@ -43,8 +43,8 @@ module.exports = {
       host: process.env.ARK_DB_HOST || 'localhost',
       port: process.env.ARK_DB_PORT || 5432,
       database: process.env.ARK_DB_DATABASE || `ark_${process.env.ARK_NETWORK_NAME}`,
-      user: process.env.ARK_DB_USERNAME || 'ark',
-      password: process.env.ARK_DB_PASSWORD || 'password'
+      user: process.env.ARK_DB_USERNAME || 'fixcrypt',
```
Comment (faustbrian) [inline packages/core/lib/config/mainnet/plugins.js]: Those changes shouldn't be committed as they are your personal configuration.

**Comment by faustbrian** [comment]
API tests are failing on this branch because https://github.com/ArkEcosystem/core/blob/develop/packages/core-api/lib/versions/2/transformers/transaction.js#L18 results in a different object then before which is missing data like an ID and Recipient.

File: packages/core-database/lib/wallet-manager.js [hunk 2]
```diff
@@ -125,7 +110,7 @@ module.exports = class WalletManager {
    * @param {void}
    */
   setByPublicKey (publicKey, wallet) {
-    this.byPublicKey = this.byPublicKey.set(publicKey, wallet)
+    this.byPublicKey = this.byPublicKey[publicKey] = wallet
```
Comment (supaiku0) [inline packages/core-database/lib/wallet-manager.js]: Same.

File: packages/core-database/lib/wallet-manager.js [hunk 3]
```diff
@@ -135,7 +120,7 @@ module.exports = class WalletManager {
    * @param {void}
    */
   setByUsername (username, wallet) {
-    this.byUsername = this.byUsername.set(username, wallet)
+    this.byUsername = this.byUsername[username] = wallet
```
Comment (supaiku0) [inline packages/core-database/lib/wallet-manager.js]: Same.

File: packages/core/lib/config/mainnet/plugins.js (hunk 1 above)
Comment (fix) [reply packages/core/lib/config/mainnet/plugins.js]: yes somehow that file slipped, pushing a revert

**Comment by fix** [comment]
yes on that aspect there is a breaking change.
- I removed all the 'fixing parts' of serialization/deserialization function, because it is mixing concern.
So now these functions are only performing clean AIP11 compliant ser/der
- the only issue is id because id calculation IS DIFFERENT than v1. so the only way to get this 'metadata' is to use the constructor.
- this is where i have no good solution, to find the right id on v1, i need to calculate signature and if different, i need to fix the deserialization (case type 1 or 3) by adding/removing recipientId
- All this to say that using the constructor is far slower in all cases.

**Comment by fix** [comment]
also another source of fail is rank.
The calculation is done ONLY at change round. Otherwise it is NOT correct in the consensus meaning.
- I have removed the rank calculation from wallet-manager for that very reason.
- I we want to enable it, we need to be SURE that this is NEVER called at turn round. However there is no such data in wallet-manager.
- The only source of reliable ranking is from database.getDelegates()

**Comment by faustbrian** [comment]
If things get removed and behaviour breaks we need to fix this behaviour so all tests go green or we can't approve and merge a PR.

A lot of things were moved or removed so they need to be substituted to provide the same behaviour as before those changes for things like `core-api`.

**Comment by fix** [comment]
That is not the point.
The test is wrong, this is my point, the rank is meaningless because ill calculated, and the test does not catch the fact this is ill calculated.

**Comment by faustbrian** [comment]
99% of the tests that are failing is because transactions have no IDs, not because of a delegate rank.

**Comment by fix** [comment]
yes i am fixing them.

**Comment by fix** [comment]
no please merge i will do it, i need a bit more time

**Comment by fix** [comment]
getting remaining tests fails i have no idea where to start

    ● API 2.0 - Blocks › POST /blocks/search › should POST a search for blocks with the exact specified reward

    expect(received).toHaveLength(length)

    Expected value to have length:
      1
    Received:
      []
    received.length:
      0

      314 |       utils.expectCollection(response)
      315 |
    > 316 |       expect(response.data.data).toHaveLength(1)
          |                                  ^
      317 |
      318 |       const block = response.data.data[0]
      319 |       utils.expectBlock(block)

      at Object.it (packages/core-api/__tests__/v2/handlers/blocks.test.js:316:34)

      ● API 2.0 - Blocks › POST /blocks/search › should POST a search for blocks with the specified reward range

    expect(received).toHaveLength(length)

    Expected value to have length:
      1
    Received:
      []
    received.length:
      0

      333 |       utils.expectCollection(response)
      334 |
    > 335 |       expect(response.data.data).toHaveLength(1)
          |                                  ^
      336 |
      337 |       const block = response.data.data[0]
      338 |       utils.expectBlock(block)

      at Object.it (packages/core-api/__tests__/v2/handlers/blocks.test.js:335:34)

**Comment by supaiku0** [comment]
@fix Travis is green and locally I also can't reproduce it. Are you sure you don't have any local changes which may cause the failure?

**Comment by fix** [comment]
Not completely green on Travis but I have more error locally.

**Comment by fix** [comment]
mmm travis is ok now, but show failed here

**Comment by fix** [comment]
restarted tests, seems its getting through

**Pull request merged**